cmd/roachtest: TestRunnerTestTimeout failed

cmd/roachtest.TestRunnerTestTimeout [failed](https://tanzanite.cluster.engflow.com/invocations/default/e5d00aad-1ffd-4298-b940-6282b28c1c4d?testReportRun=1&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY21kL3JvYWNodGVzdDpyb2FjaHRlc3RfdGVzdA==) on release-25.1.3-rc @ [abdedd3dce8c0734e8b57af59e2b0262b0a8e82b](https://github.com/cockroachdb/cockroach/commits/abdedd3dce8c0734e8b57af59e2b0262b0a8e82b):

Fatal error:

```
panic: sync: WaitGroup is reused before previous Wait has returned

```
Stack: 

```
goroutine 26018 [running]:
sync.(*WaitGroup).Wait(0xc0051cc1a0?)
	GOROOT/src/sync/waitgroup.go:118 +0x74
golang.org/x/sync/errgroup.(*Group).Wait(0xc004f2fd40)
	external/org_golang_x_sync/errgroup/errgroup.go:56 +0x25
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.Wait({0xc004f2fd40?, {0x92e76b0?, 0xc0020a0000?}, 0xc0050aa4b0?})
	pkg/util/ctxgroup/ctxgroup.go:147 +0x5c
github.com/cockroachdb/cockroach/pkg/cmd/roachtest/roachtestutil/task.(*group).WaitE(0xc00017ab00)
	pkg/cmd/roachtest/roachtestutil/task/manager.go:274 +0xc8
github.com/cockroachdb/cockroach/pkg/cmd/roachtest/roachtestutil/task.(*group).Wait(...)
	pkg/cmd/roachtest/roachtestutil/task/manager.go:266
github.com/cockroachdb/cockroach/pkg/cmd/roachtest/roachtestutil/task.(*manager).Terminate.func1()
	pkg/cmd/roachtest/roachtestutil/task/manager.go:104 +0x51
created by github.com/cockroachdb/cockroach/pkg/cmd/roachtest/roachtestutil/task.(*manager).Terminate in goroutine 25976
	pkg/cmd/roachtest/roachtestutil/task/manager.go:102 +0x89
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestRunnerTestTimeout
2025/04/07 08:12:41 test_runner.go:372: test_run_id: test_user-1744013561
2025/04/07 08:12:41 test_runner.go:533: Creating new cluster for test timeout: n0cpu4 (arch="amd64")
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=1</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/test-eng
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRunnerTestTimeout.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-49194

--- a/pkg/cmd/roachtest/roachtestutil/task/manager.go
+++ b/pkg/cmd/roachtest/roachtestutil/task/manager.go
@@ -25,6 +25,7 @@ type (
 		Tasker
 		GroupProvider
 		Terminate(*logger.Logger)
+		Cancel()
 		CompletedEvents() <-chan Event
 	}
 
@@ -105,6 +106,11 @@ func (m *manager) Terminate(l *logger.Logger) {
 	}()
 
 	WaitForChannel(doneCh, "tasks", l)
+}
+
+// Cancel will cancel all tasks started by the manager.
+func (m *manager) Cancel() {
+	m.group.cancelAll()
 }
 
 // CompletedEvents returns a channel that will receive events for all tasks

--- a/pkg/cmd/roachtest/test_runner.go
+++ b/pkg/cmd/roachtest/test_runner.go
@@ -1363,12 +1363,13 @@ func (r *testRunner) runTest(
 		// avoid situations where a test times out and the flake assignment logic fails.
 		monitorForPreemptedVMs(runCtx, t, c, l)
 
-		monitorTasks(runCtx, t.taskManager, t, l)
+		defer monitorTasks(runCtx, t.taskManager, t, l)()
 		if t.spec.Monitor {
 			testMonitor.start()
 		}
 		// This is the call to actually run the test.
 		s.Run(runCtx, t, c)
+
 	}()
 
 	var timedOut bool
@@ -1394,6 +1395,7 @@ func (r *testRunner) runTest(
 		if err := c.AddGrafanaAnnotation(ctx, t.L(), grafana.AddAnnotationRequest{Text: annotationText}); err != nil {
 			t.L().Printf(errors.Wrap(err, "error adding annotation for test end").Error())
 		}
+
 	case <-time.After(timeout):
 		// NB: We're adding the timeout failure intentionally without cancelling the context
 		// to capture as much state as possible during artifact collection.
@@ -1402,6 +1404,7 @@ func (r *testRunner) runTest(
 		// We suppress other failures from being surfaced to the top as the timeout is always going
 		// to be the main error and subsequent errors (i.e. context cancelled) add noise.
 		t.suppressFailures()
+
 		timedOut = true
 	}
 
@@ -1610,12 +1613,6 @@ func (r *testRunner) postTestAssertions(
 func (r *testRunner) teardownTest(
 	ctx context.Context, t *testImpl, c *clusterImpl, timedOut bool,
 ) error {
-	defer func() {
-		// Terminate tasks to ensure that any stray tasks are cleaned up.
-		t.L().Printf("terminating tasks")
-		t.taskManager.Terminate(t.L())
-	}()
-
 	if timedOut || t.Failed() || roachtestflags.AlwaysCollectArtifacts {
 		err := r.collectArtifacts(ctx, t, c, timedOut, time.Hour)
 		if err != nil {
@@ -1633,6 +1630,9 @@ func (r *testRunner) teardownTest(
 				t.mu.cancel()
 			}
 			t.L().Printf("test timed out; check __stacks.log and CRDB logs for goroutine dumps")
+
+			// Cancel tasks to ensure that any stray tasks are cleaned up.
+			t.taskManager.Cancel()
 		}
 		return err
 	}
@@ -2276,7 +2276,9 @@ var pollPreemptionInterval struct {
 	interval time.Duration
 }
 
-func monitorTasks(ctx context.Context, taskManager task.Manager, t test.Test, l *logger.Logger) {
+func monitorTasks(
+	ctx context.Context, taskManager task.Manager, t test.Test, l *logger.Logger,
+) func() {
 	// Monitor the task manager for completed events, or failure events and log
 	// them. A failure will call t.Errorf which cancels the test's context.
 	go func() {
@@ -2296,6 +2298,15 @@ func monitorTasks(ctx context.Context, taskManager task.Manager, t test.Test, l 
 			}
 		}
 	}()
+
+	return func() {
+		// Terminate tasks to ensure that any stray tasks are cleaned up.
+		// Tasks can only be safely terminated after the test has returned. If
+		// we terminate the manager before test code has finished executing, the
+		// test could try to initiate new tasks resulting in undefined behavior.
+		t.L().Printf("terminating stray tasks")
+		taskManager.Terminate(t.L())
+	}
 }
 
 func monitorForPreemptedVMs(ctx context.Context, t test.Test, c cluster.Cluster, l *logger.Logger) {

